modules/auxiliary/admin/dcerpc/esc_update_ldap_object.rb

```diff
@@ -0,0 +1,148 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'ruby_smb/dcerpc/client'
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::LDAP
+  include Msf::Exploit::Remote::MsIcpr
+  include Msf::Exploit::Remote::SMB::Client::Authenticated
+  include Msf::Exploit::Remote::DCERPC
+  include Msf::Auxiliary::Report
+  include Msf::OptionalSession::SMB
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Exploits AD CS Template misconfigurations which involve updating an LDAP object: ESC9 and ESC10',
+        'Description' => %q{
+          This module updates an LDAP object with a new value before requesting a certificate.
+          Request certificates via MS-ICPR (Active Directory Certificate Services). Depending on the certificate
+          template's configuration the resulting certificate can be used for various operations such as authentication.
+          PFX certificate files that are saved are encrypted with a blank password.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Will Schroeder', # original idea/research
+          'Lee Christensen', # original idea/research
+          'Oliver Lyak', # certipy implementation
+          'Spencer McIntyre', # icpr_cert module implementation
+          'jheysel-r7' # module implementation
+        ],
+        'References' => [
+          [ 'URL', 'https://github.com/GhostPack/Certify' ],
+          [ 'URL', 'https://github.com/ly4k/Certipy' ],
+          [ 'URL', 'https://medium.com/@offsecdeer/adcs-exploitation-series-part-2-certificate-mapping-esc15-6e19a6037760' ]
+        ],
+        'Notes' => {
+          'Reliability' => [],
+          'Stability' => [],
+          'SideEffects' => [ IOC_IN_LOGS ],
+          'AKA' => [ 'ESC9', 'ESC10' ]
+        },
+        'Actions' => [
+          [ 'REQUEST_CERT', { 'Description' => 'Request a certificate' } ]
+        ],
+        'DefaultAction' => 'REQUEST_CERT'
+      )
+    )
+
+    register_options([
+      OptEnum.new('UPDATE_LDAP_OBJECT', [ true, 'Either userPrincipalName or dNSHostName, Updates the necessary object of a specific user before requesting the cert. Used to exploit ESC9 and ESC10. ', 'userPrincipalName', %w[userPrincipalName dNSHostName] ]),
+      OptString.new('TARGET_USERNAME', [true, 'The username of the target LDAP object.']),
+      OptString.new('NEW_VALUE', [true, 'The new value for the specified attribute.']),
+    ])
+
+    register_advanced_options(
+      [
+        OptString.new('BASE_DN', [false, 'LDAP base DN if you already have it']),
+        OptInt.new('LDAPRport', [false, 'The target LDAP port.', 389]),
+      ]
+    )
```

Comment:
I think setting the ALT_SID, PFX and ON_BEHALF_OF options would potentially actually break the attack right? In that case, those options should probably be deregistered so users know they shouldn't be used in this context.

If they do work though we can go ahead and keep them but that's probably unnecessary because if users want that degree of control, they're probably intending to use the `icpr_cert` module and `ldap_object_attribute` modules themselves.